packages/server/lib/browsers/cdp_automation.ts

```diff
@@ -0,0 +1,130 @@
+import _ from 'lodash'
+import cdp from 'devtools-protocol'
+import tough from 'tough-cookie'
+
+interface CyCookie {
+  name: string
+  value: string
+  expirationDate: number
+  hostOnly: boolean
+  domain: string
+  path: string
+  secure: boolean
+  httpOnly: boolean
+}
+
+const cors = require('../util/cors')
+
+export function CdpAutomation (opts: {
+  invokeViaDebugger: (message: string, data?: any) => Promise<any>
+}) {
+  const { invokeViaDebugger } = opts
+
+  const normalizeGetCookieProps = function (cookie: cdp.Network.Cookie): CyCookie {
+    if (cookie.expires === -1) {
+      delete cookie.expires
+    }
+
+    // @ts-ignore
+    cookie.expirationDate = cookie.expires
+    delete cookie.expires
+
+    // @ts-ignore
+    return cookie
+  }
+
+  const normalizeGetCookies = function (cookies: cdp.Network.Cookie[]) {
+    return _.map(cookies, normalizeGetCookieProps)
+  }
+
+  const normalizeSetCookieProps = function (cookie: CyCookie): cdp.Network.SetCookieRequest {
+    cookie.name || (cookie.name = '') // name can't be undefined/null
+    cookie.value || (cookie.value = '') // ditto
+    // @ts-ignore
+    cookie.expires = cookie.expirationDate
+    if (!cookie.hostOnly && cookie.domain[0] !== '.') {
+      let parsedDomain = cors.parseDomain(cookie.domain)
+
+      // not a top-level domain (localhost, ...) or IP address
+      if (parsedDomain && parsedDomain.tld !== cookie.domain) {
+        cookie.domain = `.${cookie.domain}`
+      }
+    }
+
+    delete cookie.hostOnly
+    delete cookie.expirationDate
```

Comment:
this logic is really opaque, would be nice to comment what we're doing here and why. this is obviously handling a bunch of different situations, so either point to the tests that cover this - or try to describe it succinctly (or both)

Reply:
i'm having deja vu, i'm 99% sure i put the link to the chromium codesearch in here back when we did this the first time... let me see if i can find it again

Reply:
b2323e059